src/vm/eventpipebuffermanager.cpp

```diff
@@ -754,29 +758,38 @@ void EventPipeBufferManager::SuspendWriteEvent()
     // hadn't yet relinquished it.
     {
         SpinLockHolder _slh(&m_lock);
-        SListElem<EventPipeBufferList*> *pElem = m_pPerThreadBufferList->GetHead();
+        SListElem<EventPipeBufferList *> *pElem = m_pPerThreadBufferList->GetHead();
         while (pElem != NULL)
         {
             // Get the list and remove it from the thread.
-            EventPipeBufferList *pBufferList = pElem->GetValue();
-            for (EventPipeBuffer* pBuffer = pBufferList->GetHead(); pBuffer != nullptr; pBuffer = pBuffer->GetNext())
+            EventPipeBufferList *const pBufferList = pElem->GetValue();
+            if (pBufferList != nullptr)
             {
-                // Above we guaranteed that other threads wouldn't acquire new buffers or keep the ones they
-                // already have indefinitely, but we haven't quite guaranteed the buffer has been relinquished
-                // back to us. It's possible the WriteEvent thread allocated the buffer before we took m_lock
-                // above, but it hasn't yet acquired EventPipeThread::m_lock in order to observe that it needs
-                // to relinquish the buffer. In this state, it has a pointer to the buffer stored in registers
-                // or on the stack. If the thread is in that tiny window, all we have to do is wait for it.
-                YIELD_WHILE(pBuffer->GetVolatileState() != EventPipeBufferState::READ_ONLY);
+                EventPipeThread *const pEventPipeThread = pBufferList->GetThread();
+                if (pEventPipeThread != nullptr)
+                {
+                    YIELD_WHILE(pEventPipeThread->GetSessionWriteInProgress() == sessionId);
+                    // It still guarantees that the thread has returned its buffer, but it also now guarantees that
+                    // that the thread has returned from Session::WriteEvent() and has relinquished the session pointer
+                    // This yield is guaranteed to eventually finish because threads will eventually exit WriteEvent()
+                    // setting the flag back to -1. If the thread could quickly re-enter WriteEvent and set the flag
+                    // back to this_session_id we could theoretically get unlucky and never observe the gap, but
+                    // setting s_pSessions[this_session_id] = NULL above guaranteed that can't happen indefinately.
+                    // Sooner or later the thread is going to see the NULL value and once it does it won't store
+                    // this_session_id into the flag again.
+                    // As above I'm pretty sure this read barrier isn't necessary for correctness, but this isn't the hot
```

Comment:
'Above' again doesn't make as much sense when the context has changed. The read-barrier is also non-obvious because it happens inside the call to GetSessionWriteInProgress() now. You could just delete this last sentence.